src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp

```diff
@@ -17,91 +17,94 @@
  */
 
 #include "Creature.h"
-#include "ConfusedMovementGenerator.h"
+#include "Player.h"
 #include "PathGenerator.h"
 #include "MoveSplineInit.h"
 #include "MoveSpline.h"
-#include "Player.h"
+#include "ConfusedMovementGenerator.h"
 
 template<class T>
-void ConfusedMovementGenerator<T>::DoInitialize(T* unit)
+ConfusedMovementGenerator<T>::~ConfusedMovementGenerator()
 {
-    unit->AddUnitState(UNIT_STATE_CONFUSED);
-    unit->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_CONFUSED);
-    unit->GetPosition(i_x, i_y, i_z);
+    delete _path;
+}
 
-    if (!unit->IsAlive() || unit->IsStopped())
+template<class T>
+void ConfusedMovementGenerator<T>::DoInitialize(T* owner)
+{
+    if (!owner || !owner->IsAlive())
         return;
 
-    unit->StopMoving();
-    unit->AddUnitState(UNIT_STATE_CONFUSED_MOVE);
+    owner->AddUnitState(UNIT_STATE_CONFUSED);
+    owner->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_CONFUSED);
+    owner->StopMoving();
+
+    _timer.Reset(0);
+    owner->GetPosition(_reference.x, _reference.y, _reference.z);
 }
 
 template<class T>
-void ConfusedMovementGenerator<T>::DoReset(T* unit)
+void ConfusedMovementGenerator<T>::DoReset(T* owner)
 {
-    i_nextMoveTime.Reset(0);
-
-    if (!unit->IsAlive() || unit->IsStopped())
-        return;
-
-    unit->StopMoving();
-    unit->AddUnitState(UNIT_STATE_CONFUSED | UNIT_STATE_CONFUSED_MOVE);
+    DoInitialize(owner);
 }
 
 template<class T>
-bool ConfusedMovementGenerator<T>::DoUpdate(T* unit, uint32 diff)
+bool ConfusedMovementGenerator<T>::DoUpdate(T* owner, uint32 diff)
 {
-    if (unit->HasUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED | UNIT_STATE_DISTRACTED))
-        return true;
+    if (!owner || !owner->IsAlive())
+        return false;
 
-    if (i_nextMoveTime.Passed())
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
```

Comment:
It's not really specific to your PR and it's more like a rant than anything else but: it would be great one day to remove this concept of "casting" from the MoveGens. a casting creature should stay still, not because it's casting a spell but because its AI is saying to its legs (MoveGens) "do not move". The legs (the MoveGens) shouldn't know the concept of "cast" or "spell".

In concrete terms, this means that when the AI starts to cast a spell, it should turn on a unit state (or a movementflag ?) like UNIT_STATE_NOT_MOVING_BY_AI that should be included in the UNIT_STATE_NOT_MOVE mask.